### DISCUSSION TITLE
CoreWCF v1.6.0

### DISCUSSION THREAD

**Discussion by mconnew** [CoreWCF/CoreWCF#1447] [Releases]
# Release Notes

## Package Info
[CoreWCF.ConfigurationManager 1.6.0](https://www.nuget.org/packages/CoreWCF.ConfigurationManager/1.6.0)  
[CoreWCF.Http 1.6.0](https://www.nuget.org/packages/CoreWCF.Http/1.6.0)  
[CoreWCF.Kafka 1.6.0](https://www.nuget.org/packages/CoreWCF.Kafka/1.6.0)  
[CoreWCF.Kafka.Client 1.6.0](https://www.nuget.org/packages/CoreWCF.Kafka.Client/1.6.0)  
[CoreWCF.NetFramingBase 1.6.0](https://www.nuget.org/packages/CoreWCF.NetFramingBase/1.6.0)  
[CoreWCF.NetTcp 1.6.0](https://www.nuget.org/packages/CoreWCF.NetTcp/1.6.0)  
[CoreWCF.Primitives 1.6.0](https://www.nuget.org/packages/CoreWCF.Primitives/1.6.0)  
[CoreWCF.Queue 1.6.0](https://www.nuget.org/packages/CoreWCF.Queue/1.6.0)  
[CoreWCF.RabbitMQ 1.6.0](https://www.nuget.org/packages/CoreWCF.RabbitMQ/1.6.0)  
[CoreWCF.RabbitMQ.Client 1.6.0](https://www.nuget.org/packages/CoreWCF.RabbitMQ.Client/1.6.0)  
[CoreWCF.Templates 1.6.0](https://www.nuget.org/packages/CoreWCF.Templates/1.6.0)  
[CoreWCF.UnixDomainSocket 1.6.0](https://www.nuget.org/packages/CoreWCF.UnixDomainSocket/1.6.0)  
[CoreWCF.WebHttp 1.6.0](https://www.nuget.org/packages/CoreWCF.WebHttp/1.6.0)  

## .NET Compatibility
This release depends on .NET Standard 2.0 and runs on any .NET version which supports .NET Standard 2.0. This release supports .NET Framework 4.6.2 and above, .NET 6, and .NET 8. It is built on top of ASP.NET Core and has been tested and runs on all currently supported versions of ASP.NET Core up to 8.0. The CoreWCF.RabbitMQ.Client, CoreWCF.Kafka.Client, and CoreWCF.UnixDomainSocket packages only support .NET 6 or later. Support for .NET 7 has been dropped as it is now out of support.

### OperationContract initiating and terminating support
The `OperationContractAttribute` properties `IsInitiating` and `IsTerminating` have been made available by @afifi-ins in https://github.com/CoreWCF/CoreWCF/pull/1332  

### Support for injecting ServiceBehavior by service type
You can now inject a service behavior into DI to be applied to a single service type. This was added by @g7ed6e in https://github.com/CoreWCF/CoreWCF/pull/1040. You can read more about this in the [blog post published here](https://corewcf.github.io/blog/2024/08/23/keyed-servicebehavior).  

### Support for injecting message properties into operation implementation
Using operation injection support, you can now automatically inject a message property into your service operation. This was added by @g7ed6e in https://github.com/CoreWCF/CoreWCF/pull/1286. You can read more about this in the [blog post published here](https://corewcf.github.io/blog/2024/08/23/message-property-injection). We did make a couple of minor binary breaking changes with this feature, but you shouldn't see any compilation breaks. We changed some static `Name` properties on message property types to be `const string`. Static properties can't be used as values for attribute properties whereas constants can. For example, `HttpRequestMessagePropert.Name` used to be declared as `public static string Name { get; }`, and now is declared as `public const string Name;`. While this doesn't break compilation, it isn't binary compatible.  

### CoreWCF project template improvements
* Add IISExpress launch profile in templates by @g7ed6e in https://github.com/CoreWCF/CoreWCF/pull/1190
* Dropped net7.0 from templates by @g7ed6e in https://github.com/CoreWCF/CoreWCF/pull/1418  

### IOperationInvoker code generation
As part of an eventual goal to support aot for CoreWCF, we have introduced a code generator for implementing the default `IOperationInvoker`. This is an opt-in feature which you enable by setting the following in your project:
```xml
  <PropertyGroup>
    <EnableCoreWCFOperationInvokerGenerator>true</EnableCoreWCFOperationInvokerGenerator>
  </PropertyGroup>
```
This isn't sufficient to enable aot as we have other code which requires use of Ref.Emit, but is a step towards that goal.  

### New NetTcp configuration api's
As part of creating the full fix for https://github.com/CoreWCF/CoreWCF/security/advisories/GHSA-32jq-mv89-5rx7, the mechanism to configure some of the properties for NetTcp needed to be updated. You can read more about this in the [blog post published here](https://corewcf.github.io/blog/2024/08/23/new-nettcp-apis). Existing api's will continue to work so there shouldn't be any breaking changes for existing code. 

### Dependency updates
As part of the work to support keyed service dependency injection, we needed to update a few package dependencies to the .NET 8.0 versions. We run all our tests on .NET Framework 4.7.2 and .NET 6.0 so don't expect this to cause problems. If you do experience any issues due to CoreWCF upgrading dependencies, please open an issue explaining the error you're experiencing. The dependencies which were updated to 8.x versions are:
* Microsoft.Bcl.AsyncInterfaces
* Microsoft.Extensions.DependencyInjection
* Microsoft.Extensions.DependencyInjection.Abstractions
* Microsoft.Extensions.Primitives
In addition, some other packages including some logging packages were updated from 2.x/4.x versions to 6.x versions. This was due to incompatibilities between the ASP.NET Core 2.1 version of Kestrel (as used with .NET Framework) and Microsoft.Extensions.Primitives which broke logging. Updating the logging packages to 6.x fixes this and is compatible with Kestrel 2.1, later versions of Kestrel shipped with the runtime, and the later Microsoft.Extensions.Primitives package.

### Bug Fixes
* Add KafkaMessageProperty by @g7ed6e in https://github.com/CoreWCF/CoreWCF/pull/1277
* Prevent null reference exception when a client makes request to wrong endpoint address (#1331) by @bjornbouetsmith in https://github.com/CoreWCF/CoreWCF/pull/1335
* Detect kafka client side timeout by @g7ed6e in https://github.com/CoreWCF/CoreWCF/pull/1287
* Add support for binding namespace for WCF xml based configs by @VitahaPo in https://github.com/CoreWCF/CoreWCF/pull/1359
* Fix service endpoint URL in WSDL for WCF deployed in container by @VitahaPo in https://github.com/CoreWCF/CoreWCF/pull/1358
* Include CoreWCFWebApplicationExtensions.cs only within c# projects by @g7ed6e in https://github.com/CoreWCF/CoreWCF/pull/1371
* Fix NetFramingBase based transport Abort code path by @mconnew in https://github.com/CoreWCF/CoreWCF/pull/1348
* Fixed operation invoker expression based implementation to handle omitted primitive values by @g7ed6e in https://github.com/CoreWCF/CoreWCF/pull/1386
* Add multiple Kafka topic subscription using regex support by @g7ed6e in https://github.com/CoreWCF/CoreWCF/pull/1290
* Enabled MessageCredentialType.IssuedToken for WSHttpBinding by @emre-toraman in https://github.com/CoreWCF/CoreWCF/pull/1324
* Fix WSDL endpoint URLs for multiple services by @a-tze in https://github.com/CoreWCF/CoreWCF/pull/1409
* Remove NegotiateInputLengthDecider from CoreWCF.NetFramingBase which broke large messages by @SaxoAnders in https://github.com/CoreWCF/CoreWCF/pull/1424
* Make public CustomUserNameSecurityTokenAuthenticator by @lucashigor in https://github.com/CoreWCF/CoreWCF/pull/1419
* Add missing UDS ProtectionLevel property by @mconnew in https://github.com/CoreWCF/CoreWCF/pull/1433
* Add missing api's to DataContractSerializerOperationBehavior to enable configuring a resolver and/or surrogate provider by @imcarolwang in https://github.com/CoreWCF/CoreWCF/pull/1428
* Fix ByteStreamMessage behavior when calling CreateBufferedCopy and GetReaderAtBodyContents by @mconnew in https://github.com/CoreWCF/CoreWCF/pull/1436
* Fix multiple problems with listen uri being propagated when building dispatcher pipeline by @mconnew in https://github.com/CoreWCF/CoreWCF/pull/1435
* Add keyed services injection support in source gen by @g7ed6e in https://github.com/CoreWCF/CoreWCF/pull/1443

### Feedback
Your feedback is important and appreciated. Please use the discussion https://github.com/CoreWCF/CoreWCF/discussions/1447 for your questions and comments.

**Full Changelog**: https://github.com/CoreWCF/CoreWCF/compare/v1.5.2...v1.6.0

## New Contributors
* @bjornbouetsmith made their first contribution in https://github.com/CoreWCF/CoreWCF/pull/1335
* @VitahaPo made their first contribution in https://github.com/CoreWCF/CoreWCF/pull/1359
* @emre-toraman made their first contribution in https://github.com/CoreWCF/CoreWCF/pull/1324
* @a-tze made their first contribution in https://github.com/CoreWCF/CoreWCF/pull/1409
* @241 made their first contribution in https://github.com/CoreWCF/CoreWCF/pull/1420
* @SaxoAnders made their first contribution in https://github.com/CoreWCF/CoreWCF/pull/1424
* @lucashigor made their first contribution in https://github.com/CoreWCF/CoreWCF/pull/1419

**Full Changelog**: https://github.com/CoreWCF/CoreWCF/compare/v1.5.2...v1.6.0

<hr /><em>This discussion was created from the release <a href='https://github.com/CoreWCF/CoreWCF/releases/tag/v1.6.0'>CoreWCF v1.6.0</a>.</em>

**Comment by cobalto** [comment]
Thank you for the new release 🙏😌
Just to confirm, will there be no  __CoreWCF.NetNamedPipe__ `1.6.0-preview1` this time?

**Reply by mconnew** [reply]
Not this time. It's a lot of extra work to release a preview release and nothing of consequence changes in NetNamedPipe. I hope to circle back and complete NetNamedPipe work soon as it's not a lot to get it out of preview.

**Comment by hello-ldf** [comment]
Thank you all for your hard work. Looking forward to the support for reliable sessions!
